csi/node_server.go

```diff
@@ -207,6 +177,24 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 		}
 	}
 
+	if isBlock {
+		devicePath := getStageBlockVolumePath(stagingTargetPath, volumeID)
+		_, err := os.Stat(devicePath)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, status.Errorf(codes.Internal, errors.Wrapf(err, "failed to stat device %s", devicePath).Error())
+			}
+		}
+
```

Comment:
I removed the endpoint fallback like:

```
// Fall back to the controller endpoint if the device path under the stagingTargetPath doesn't exist
log.Infof("Device path %s doesn't exist, falling back to controller endpoint %s", devicePath, volume.Controllers[0].Endpoint)
devicePath = volume.Controllers[0].Endpoint
```

This was necessary before because `NodeStageVolume` was a no-op in < `v1.6.0`, so a volume might not be staged at `stagingTargetPath` before `NodePublishVolume` is called. The new restaging logic will ensure all `NodePublishVolume` commands succeed (with automatic restaging if necessary). Even if a volume was previously staged (via no-op) in < `v1.6.0`, calling `NodeUnstageVolume` and `NodeRestageVolume` is safe to do. See the new comment in `restageRequired` for additional details.